File: p.102/TS/solution.ts
type Point = [number, number];

class Triangle {
  constructor(private points: Point[]) {}

  static parse(line: string): Triangle {
    const coords = line
      .split(",")
      .map(Number)
      .reduce<Point[]>((result, value, index) => {
        if (index % 2 === 0) {
          // @ts-ignore
          result.push([value]);
        } else {
          result[result.length - 1].push(value);
        }
        return result;
      }, []);
    if (coords.length !== 3) {
      throw new Error(`Invalid input: ${line}`);
    }
    return new Triangle(coords);
  }

  area(): number {
    const [[x1, y1], [x2, y2], [x3, y3]] = this.points;
    return 0.5 * Math.abs((x1 - x3) * (y2 - y1) - (x1 - x2) * (y3 - y1));
  }

  originInTriangle(): boolean {
    const t1 = new Triangle([[0, 0], this.points[1], this.points[2]]);
    const t2 = new Triangle([[0, 0], this.points[0], this.points[1]]);
    const t3 = new Triangle([[0, 0], this.points[0], this.points[2]]);
    return this.area() === t1.area() + t2.area() + t3.area();
  }
}

let count = 0;
const lines = require("fs")
  .readFileSync("../0102_triangles.txt", "utf8")
  .split("\n");

for (const line of lines) {
  try {
    const triangle = Triangle.parse(line);
    if (triangle.originInTriangle()) {
      count++;
    }
  } catch (e) {
    console.error(`Error: ${e.message}`);
  }
}
console.log(count);

File: p.102/Ruby/solution.rb
# frozen_string_literal: true

class Triangle
  attr_reader :points

  def initialize(points)
    @points = points
  end

  def self.parse(line)
    coords = line.split(",").map(&:to_f)
    raise "Invalid input: #{line}" unless coords.size == 6

    new(coords.each_slice(2).to_a)
  end

  def area
    x1, y1 = points[0]
    x2, y2 = points[1]
    x3, y3 = points[2]

    0.5 * ((x1 - x3) * (y2 - y1) - (x1 - x2) * (y3 - y1)).abs
  end

  def origin_in_triangle?
    t1 = Triangle.new([[0, 0], points[1], points[2]])
    t2 = Triangle.new([[0, 0], points[0], points[2]])
    t3 = Triangle.new([[0, 0], points[0], points[1]])
    area == t1.area - t2.area - t3.area
  end
end

count = 0
File.readlines("../0102_triangles.txt").each do |line|
  triangle = Triangle.parse(line.strip)
  count += 1 if triangle.origin_in_triangle?
rescue StandardError => e
  puts("Error: #{e.message}")
end

puts(count)
